### DISCUSSION TITLE
Error predicting text data using LSTM model (ML Capstone Project)

### DISCUSSION THREAD

**Discussion by brigittacarmen** [teamalgoritma/community#746] [Q&A] [answered]
Dear team Algoritma,

Saya mempunyai masalah dengan error di tahap predict menggunakan model LSTM.

Berikut saya lampirkan tahapan-tahapan yang telah saya kerjakan beserta errornya:

<img width="454" alt="image" src="https://github.com/teamalgoritma/community/assets/134923285/0156dfc9-529a-405a-a9f2-f1a6136a9dc8">

<img width="470" alt="image" src="https://github.com/teamalgoritma/community/assets/134923285/6cfe41c2-5175-43ec-983a-3a6eeb3abd0c">

<img width="446" alt="image" src="https://github.com/teamalgoritma/community/assets/134923285/d8db7c01-648e-4c0a-b275-f758207508d4">

Dan dibawah ini, ss bahwa model telah dibuat 

<img width="458" alt="image" src="https://github.com/teamalgoritma/community/assets/134923285/814d24e6-28a9-4d2f-b7af-b05e79815416">

Lalu melewati proses fitting yang berhasil juga

<img width="449" alt="image" src="https://github.com/teamalgoritma/community/assets/134923285/3db57aea-3960-4b1e-a752-1abcb15ea72b">

Tetapi saat saya predict, keluar error seperti ini, sementara data_val_x sudah saya jadikan as.String.

<img width="455" alt="image" src="https://github.com/teamalgoritma/community/assets/134923285/9c8383ef-8402-4a7f-9a06-0c2d5ca39028">

Apakah ada jalan keluarnya?
Any response will be highly appreciated. Terima kasih!

Brigitta Carmen

**Comment by kevbow** [comment]
Dear @brigittacarmen,

Fungsi `predict_classes()` sudah tidak dapat digunakan pada versi tensorflow yang Anda gunakan. Silakan coba gunakan kode berikut seperti yang telah diajarkan di kelas:
```r
val_pred <- predict(model_jadi, data_val_x) %>% 
  k_argmax() %>%
  as.array() %>% 
  as.factor()
```

Semoga membantu, terima kasih.

**Reply by brigittacarmen** [reply]
Terima kasih Mas Kevin atas responsenya.

Sudah saya coba jalankan, tetapi masih mengeluarkan error.
Saya juga sudah mencoba cara yang kedua (tertera di screenshot) tetapi masih mengeluarkan error yang sama juga.

<img width="482" alt="image" src="https://github.com/teamalgoritma/community/assets/134923285/a2c7a3ae-83ec-48dd-ba0c-9e7d21e3eeba">

**Reply by kevbow** [reply, marked as answer]
Setelah saya perhatikan, Anda melakukan save history model ke dalam bentuk `.rds`. Pada model tensorflow sebaiknya Anda menyimpan modelnya menggunakan fungsi `save_model_tf()` dan variable yang di-save bukan `history` melainkan `model`.
Apakah jika menggunakan variable `model` (object model yang di-training) predict dapat dilakukan?
Silakan coba kode berikut (tentunya setelah melakukan fitting model ulang):
```r
model %>% predict(data_val_x)
```

Dan gunakanlah kode berikut untuk save dan load model
```r
# save model
save_model_tf(model, "model_lstm") # pastikan yang dimasukkan adalah model bukan history

# load model
lstm_train <- load_model_tf("model_lstm")

# pengecekan
summary(lstm_train)
```

*sudah diedit dengan kode yang sesuai.

**Reply by brigittacarmen** [reply]
Sudah dicoba Mas, dan berhasil untuk mempredict.

![image](https://github.com/teamalgoritma/community/assets/134923285/14f2d373-52b4-42c6-ad8d-07a157d14325)

Untuk hasil prediksinya, saya bingung cara mengubah menjadi spam/hamnya.

Sementara untuk savenya masih mengeluarkan error juga

![image](https://github.com/teamalgoritma/community/assets/134923285/d21256b7-a3b2-4693-90bf-74c1ecf525c5)

**Reply by kevbow** [reply]
Mohon maaf ada kesalahan penulisan pada kode yang sebelumnya saya berikan. Berikut saya perbaiki pada bagian load model-nya.
```r
# load model
lstm_train <- load_model_tf("model_lstm")

# pengecekan
summary(lstm_train)
```

**Comment by brigittacarmen** [comment]
Terima kasih atas jawabannya,

Mohon maaf izin memastikan lagi mas, untuk predict() nya apakah masih tetap harus menggunakan k_argmax(), as.array() dan as.factor()?

**Reply by kevbow** [reply]
Betul Bu, karena output dari model Ibu adalah peluang dari masing-masing kelas maka perlu menggunakan argmax (mengambil kelas yang peluangnya paling maksimal/tinggi).

**Reply by brigittacarmen** [reply]
Baik terima kasih banyak Mas
